api/admin/controller/catalog_services.py

```diff
@@ -67,24 +73,65 @@ def process_post(self):
             service.name = name
 
         [protocol] = [p for p in self.protocols if p.get("name") == protocol]
+
         result = self._set_integration_settings_and_libraries(service, protocol)
         if isinstance(result, ProblemDetail):
             return result
 
+        external_integration_link = self._set_external_integration_link(service)
+        if isinstance(external_integration_link, ProblemDetail):
+            return external_integration_link
+
         library_error = self.check_libraries(service)
         if library_error:
             self._db.rollback()
             return library_error
 
-        storage_protocol_error = self.check_storage_protocol(service)
-        if storage_protocol_error:
-            self._db.rollback()
-            return storage_protocol_error
-
         if is_new:
             return Response(unicode(service.id), 201)
         else:
             return Response(unicode(service.id), 200)
+    
+    def _set_external_integration_link(self, service):
+        mirror_integration_id = flask.request.form.get('mirror_integration_id')
+        if mirror_integration_id == self.NO_MIRROR_INTEGRATION:
+            current_integration_link = get_one(
+                self._db,
+                ExternalIntegrationLink,
+                library_id=None,
+                external_integration_id=service.id,
+                purpose="MARC"
+            )
+            if current_integration_link:
+                self._db.delete(current_integration_link)
+        else:
+            storage_integration = get_one(
+                self._db, ExternalIntegration, id=mirror_integration_id
+            )
+            # set_trace()
+            if not storage_integration or not storage_integration.setting(S3Uploader.MARC_BUCKET_KEY).value:
+                return MISSING_INTEGRATION
+            self._update_external_integration_link(
+                self._db,
+                external_integration=service,
+                other_external_integration=storage_integration
+            )
+
+    def _update_external_integration_link(
+            self, _db, external_integration, other_external_integration
+    ):
+        """Find or create a ExternalIntegrationLink associated with a Library
+        and an ExternalIntegration.
+        """
+        external_integration_link, ignore = get_one_or_create(
+            _db, ExternalIntegrationLink,
+            library_id=None,
+            external_integration_id=external_integration.id,
+            purpose="MARC"
+        )
+
+        external_integration_link.other_integration_id=other_external_integration.id
+        _db.commit()
```

Comment:
I don't think you need to commit here -- it's pretty rare. Try it without the `commit` and see if it works.